### ISSUE TITLE
Spectral regularization confusion

### ISSUE THREAD

**Issue by a-gardner1** [alanbacellar/DWN#4]
Thanks for providing this repo for your paper! I'm exploring weightless neural networks for one of my own projects, and I can't seem to identify anywhere in the code that implements the spectral regularization mentioned in the paper.

In particular, I don't see how the regularization as described in the paper works unless I am misreading something or there is a typo. Based on my understanding of Parseval's theorem (O'Donnell, 2021), the L2 norm of the Fourier coefficients of a function $f : \\\{-1, 1\\\}^n \to \\\{-1, 1\\\}$ should always be equal to one. Since the LUTs described in the paper describe such functions, $\text{specnorm}(L) = u$ regardless of whatever the data matrix $L$ contains. This matches my own observations after implementing the normalization according to what should be the equivalent formulation below:

```math
C^\intercal_{ij} = \prod_{k=0}^{n-1} \left[\left(2\left(\frac{j}{2^k}\mod{2}\right) - 1\right)\left(\frac{i}{2^k}\mod{2}\right) + 1 - \left(\frac{i}{2^k}\mod{2}\right)\right]
```

I'd greatly appreciate a response as soon as is convenient. Thank you

For reference, here is my test script, where $M=L^\intercal$ and $P = C^\intercal$:

```python
import torch
max_n = 10
norms = []
alt_norms = []
for n in range(max_n):
    m = 2**n
    w = 1
    M = 2*(torch.randn(m, w) > 0).float() - 1
    P = torch.zeros(m,m, dtype=torch.long)
    for i in range(m):
        for j in range(m):
            val = 1
            for k in range(0, n):
                val *= (2*((j >> k) % 2) - 1)*((i>>k) % 2) + 1 - ((i>>k) % 2)
            P[i,j] = val
    P = P.to(torch.float)
    print(torch.matmul(P.T, P))
    print(P)
    norms.append(torch.matmul(P, M).norm())
    alt_norms.append(torch.matmul(M.T, torch.matmul(torch.matmul(P.T, P), M)).sum().sqrt())
print(norms)
print(alt_norms)
```
